Refactored ops on size to be dispatcher ops

Chillee opened the pull request:
An example of how the graph looks now.
```
def forward(self, x_1):
    size = torch.ops.math.size(x_1, 0)
    size_1 = torch.ops.math.size(x_1, 1);  x_1 = None
    ones = torch.ops.aten.ones.default([1], device = device(type='cpu'), pin_memory = False)
    expand_sym_int = torch.ops.aten.expand.SymInt(ones, [size, size_1]);  ones = size = size_1 = None
    cos_default = torch.ops.aten.cos.default(expand_sym_int);  expand_sym_int = None
    return (cos_default,)
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83813
* #83814
* #83822
* __->__ #83719

Chillee committed:
Refactored ops on size to be dispatcher ops

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -18,6 +18,23 @@
+math_lib.define("mul(int a, int b) -> int")
+
+def size(x, dim):
+    return x.size(dim)
```
Why aren't these just the aten.size ops?

Chillee replied:
that's a ... good point lol.

Chillee committed:
Update on "Refactored ops on size to be dispatcher ops"

An example of how the graph looks now.
```
def forward(self, x_1):
    size = torch.ops.math.size(x_1, 0)
    size_1 = torch.ops.math.size(x_1, 1);  x_1 = None
    ones = torch.ops.aten.ones.default([1], device = device(type='cpu'), pin_memory = False)
    expand_sym_int = torch.ops.aten.expand.SymInt(ones, [size, size_1]);  ones = size = size_1 = None
    cos_default = torch.ops.aten.cos.default(expand_sym_int);  expand_sym_int = None
    return (cos_default,)
```




[ghstack-poisoned]

Chillee committed:
Update on "Refactored ops on size to be dispatcher ops"

An example of how the graph looks now.
```
def forward(self, x_1):
    size = torch.ops.math.size(x_1, 0)
    size_1 = torch.ops.math.size(x_1, 1);  x_1 = None
    ones = torch.ops.aten.ones.default([1], device = device(type='cpu'), pin_memory = False)
    expand_sym_int = torch.ops.aten.expand.SymInt(ones, [size, size_1]);  ones = size = size_1 = None
    cos_default = torch.ops.aten.cos.default(expand_sym_int);  expand_sym_int = None
    return (cos_default,)
```




[ghstack-poisoned]

Chillee committed:
Update on "Refactored ops on size to be dispatcher ops"

An example of how the graph looks now.
```
def forward(self, x_1):
    size = torch.ops.math.size(x_1, 0)
    size_1 = torch.ops.math.size(x_1, 1);  x_1 = None
    ones = torch.ops.aten.ones.default([1], device = device(type='cpu'), pin_memory = False)
    expand_sym_int = torch.ops.aten.expand.SymInt(ones, [size, size_1]);  ones = size = size_1 = None
    cos_default = torch.ops.aten.cos.default(expand_sym_int);  expand_sym_int = None
    return (cos_default,)
```




[ghstack-poisoned]

Chillee committed:
Update on "Refactored ops on size to be dispatcher ops"

An example of how the graph looks now.
```
def forward(self, x_1):
    size = torch.ops.math.size(x_1, 0)
    size_1 = torch.ops.math.size(x_1, 1);  x_1 = None
    ones = torch.ops.aten.ones.default([1], device = device(type='cpu'), pin_memory = False)
    expand_sym_int = torch.ops.aten.expand.SymInt(ones, [size, size_1]);  ones = size = size_1 = None
    cos_default = torch.ops.aten.cos.default(expand_sym_int);  expand_sym_int = None
    return (cos_default,)
```




[ghstack-poisoned]

Chillee committed:
Update on "Refactored ops on size to be dispatcher ops"

An example of how the graph looks now.
```
def forward(self, x_1):
    size = torch.ops.math.size(x_1, 0)
    size_1 = torch.ops.math.size(x_1, 1);  x_1 = None
    ones = torch.ops.aten.ones.default([1], device = device(type='cpu'), pin_memory = False)
    expand_sym_int = torch.ops.aten.expand.SymInt(ones, [size, size_1]);  ones = size = size_1 = None
    cos_default = torch.ops.aten.cos.default(expand_sym_int);  expand_sym_int = None
    return (cos_default,)
```




[ghstack-poisoned]

Chillee committed:
Update on "Refactored ops on size to be dispatcher ops"

An example of how the graph looks now.
```
def forward(self, x_1):
    size = torch.ops.math.size(x_1, 0)
    size_1 = torch.ops.math.size(x_1, 1);  x_1 = None
    ones = torch.ops.aten.ones.default([1], device = device(type='cpu'), pin_memory = False)
    expand_sym_int = torch.ops.aten.expand.SymInt(ones, [size, size_1]);  ones = size = size_1 = None
    cos_default = torch.ops.aten.cos.default(expand_sym_int);  expand_sym_int = None
    return (cos_default,)
```




[ghstack-poisoned]

Chillee committed:
Update on "Refactored ops on size to be dispatcher ops"

An example of how the graph looks now.
```
def forward(self, x_1):
    size = torch.ops.math.size(x_1, 0)
    size_1 = torch.ops.math.size(x_1, 1);  x_1 = None
    ones = torch.ops.aten.ones.default([1], device = device(type='cpu'), pin_memory = False)
    expand_sym_int = torch.ops.aten.expand.SymInt(ones, [size, size_1]);  ones = size = size_1 = None
    cos_default = torch.ops.aten.cos.default(expand_sym_int);  expand_sym_int = None
    return (cos_default,)
```




[ghstack-poisoned]

Chillee committed:
Update on "Refactored ops on size to be dispatcher ops"

An example of how the graph looks now.
```
def forward(self, x_1):
    size = torch.ops.math.size(x_1, 0)
    size_1 = torch.ops.math.size(x_1, 1);  x_1 = None
    ones = torch.ops.aten.ones.default([1], device = device(type='cpu'), pin_memory = False)
    expand_sym_int = torch.ops.aten.expand.SymInt(ones, [size, size_1]);  ones = size = size_1 = None
    cos_default = torch.ops.aten.cos.default(expand_sym_int);  expand_sym_int = None
    return (cos_default,)
```




[ghstack-poisoned]

Chillee committed:
Update on "Refactored ops on size to be dispatcher ops"

An example of how the graph looks now.
```
def forward(self, x_1):
    size = torch.ops.math.size(x_1, 0)
    size_1 = torch.ops.math.size(x_1, 1);  x_1 = None
    ones = torch.ops.aten.ones.default([1], device = device(type='cpu'), pin_memory = False)
    expand_sym_int = torch.ops.aten.expand.SymInt(ones, [size, size_1]);  ones = size = size_1 = None
    cos_default = torch.ops.aten.cos.default(expand_sym_int);  expand_sym_int = None
    return (cos_default,)
```




[ghstack-poisoned]

Chillee committed:
Update on "Refactored ops on size to be dispatcher ops"

An example of how the graph looks now.
```
def forward(self, x_1):
    size = torch.ops.math.size(x_1, 0)
    size_1 = torch.ops.math.size(x_1, 1);  x_1 = None
    ones = torch.ops.aten.ones.default([1], device = device(type='cpu'), pin_memory = False)
    expand_sym_int = torch.ops.aten.expand.SymInt(ones, [size, size_1]);  ones = size = size_1 = None
    cos_default = torch.ops.aten.cos.default(expand_sym_int);  expand_sym_int = None
    return (cos_default,)
```




[ghstack-poisoned]

Chillee committed:
Update on "Refactored ops on size to be dispatcher ops"

An example of how the graph looks now.
```
def forward(self, x_1):
    size = torch.ops.math.size(x_1, 0)
    size_1 = torch.ops.math.size(x_1, 1);  x_1 = None
    ones = torch.ops.aten.ones.default([1], device = device(type='cpu'), pin_memory = False)
    expand_sym_int = torch.ops.aten.expand.SymInt(ones, [size, size_1]);  ones = size = size_1 = None
    cos_default = torch.ops.aten.cos.default(expand_sym_int);  expand_sym_int = None
    return (cos_default,)
```




[ghstack-poisoned]

Chillee committed:
Update on "Refactored ops on size to be dispatcher ops"

An example of how the graph looks now.
```
def forward(self, x_1):
    size = torch.ops.math.size(x_1, 0)
    size_1 = torch.ops.math.size(x_1, 1);  x_1 = None
    ones = torch.ops.aten.ones.default([1], device = device(type='cpu'), pin_memory = False)
    expand_sym_int = torch.ops.aten.expand.SymInt(ones, [size, size_1]);  ones = size = size_1 = None
    cos_default = torch.ops.aten.cos.default(expand_sym_int);  expand_sym_int = None
    return (cos_default,)
```




[ghstack-poisoned]

ezyang commented on test/test_proxy_tensor.py:
```diff
@@ -710,10 +710,10 @@ def f(a):
-    size = a_1.size(0);  a_1 = None
-    mul = size * 2;  size = None
+    size = torch.ops.aten.size(a_1, 0);  a_1 = None
+    mul = torch.ops.math.mul(size, 2);  size = None
```
good bye size * 2, I will miss you lol

Chillee replied:
It's back :)

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -566,24 +566,20 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        flat_symints = [
+            i
+            for i in tree_flatten((args, kwargs))[0]
+            if isinstance(i, torch._C.SymIntNode)
```
can someone pleaaase give this a better name lol. torch.SymIntNode or torch.SymInt

Chillee replied:
where do I register it to give it a better name?

ezyang replied:
I think monkeying around with `torch/__init__.py` so it gets set there is the idea

albanD replied:
If you want printing to also look nice, you can update the `__module__` field to match.

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -467,7 +468,18 @@ def __sym_dispatch__(self, func, types, args, kwargs):
+        }
+        if func not in mapped:
+            print(func)
+            raise RuntimeError(f"Don't support {func} on symints")
```
This should get fixed at the source. SymDispatchMode should get torch.ops.math calls to begin with. Don't post facto translate them in proxy tensor.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -18,6 +18,19 @@
+math_lib.define("mul(int a, int b) -> int")
+math_lib.define("eq(int a, int b) -> bool")
+math_lib.define("gt(int a, int b) -> bool")
+math_lib.define("lt(int a, int b) -> bool")
```
FWIW, JIT "prim" actually already defines operators that look like these. They don't have dispatcher entries though.

ezyang replied:
How come these don't have SymInt in their schema?

albanD commented on functorch/functorch/_src/fx_minifier.py:
```diff
@@ -296,4 +296,5 @@ def try_granularity(failing_state, granularity, use_non_granular):
     print(f"Made {num_queries} queries")
     failing_fx = fx.GraphModule(fail_f, failing_state.graph)
     dump_state(failing_fx, failing_state.inps)
+    print("Wrote minimal repro out to repro.py")
```
raw print?

Chillee replied:
No, this is intended - just a small QOL improvement to minifier I'm folding into this change lol.

bdhirsh commented on torch/_meta_registrations.py:
```diff
@@ -211,6 +211,23 @@ def meta_bernoulli(self, *, generator=None, out):
+    layout = self.layout if layout is None else layout
+    assert pin_memory is None
+    assert memory_format is None
+    return self.new_empty(self.shape, dtype=dtype, device=device, pin_memory=pin_memory)
```
just curious - is this python decomp needed now? (aka something now breaks with the C++ decomp)

Chillee replied:
`_to_copy` isn't `CompositeImplicitAutograd`, so I don't think there's a "C++ decomp"? Or do you mean the meta function implemented in C++?

bdhirsh replied:
oh yep, I mean the meta function implemented in C++; is that not useable with the changes in this PR?

Chillee replied:
I don't think we use any C++ meta functions today, although I think it's plausible we could in the future.

IvanYashchuk replied:
This implementation is slightly incorrect and it seems there were no tests to catch that: https://github.com/pytorch/pytorch/issues/84335.

Chillee committed:
Update on "Refactored ops on size to be dispatcher ops"

An example of how the graph looks now.
```
def forward(self, x_1):
    size = torch.ops.math.size(x_1, 0)
    size_1 = torch.ops.math.size(x_1, 1);  x_1 = None
    ones = torch.ops.aten.ones.default([1], device = device(type='cpu'), pin_memory = False)
    expand_sym_int = torch.ops.aten.expand.SymInt(ones, [size, size_1]);  ones = size = size_1 = None
    cos_default = torch.ops.aten.cos.default(expand_sym_int);  expand_sym_int = None
    return (cos_default,)
```




[ghstack-poisoned]

Chillee committed:
Update on "Refactored ops on size to be dispatcher ops"

An example of how the graph looks now.
```
def forward(self, x_1):
    size = torch.ops.math.size(x_1, 0)
    size_1 = torch.ops.math.size(x_1, 1);  x_1 = None
    ones = torch.ops.aten.ones.default([1], device = device(type='cpu'), pin_memory = False)
    expand_sym_int = torch.ops.aten.expand.SymInt(ones, [size, size_1]);  ones = size = size_1 = None
    cos_default = torch.ops.aten.cos.default(expand_sym_int);  expand_sym_int = None
    return (cos_default,)
```




[ghstack-poisoned]

Chillee committed:
Update on "Refactored ops on size to be dispatcher ops"

An example of how the graph looks now.
```
def forward(self, x_1):
    size = torch.ops.math.size(x_1, 0)
    size_1 = torch.ops.math.size(x_1, 1);  x_1 = None
    ones = torch.ops.aten.ones.default([1], device = device(type='cpu'), pin_memory = False)
    expand_sym_int = torch.ops.aten.expand.SymInt(ones, [size, size_1]);  ones = size = size_1 = None
    cos_default = torch.ops.aten.cos.default(expand_sym_int);  expand_sym_int = None
    return (cos_default,)
```




[ghstack-poisoned]

Chillee committed:
Update on "Refactored ops on size to be dispatcher ops"

An example of how the graph looks now.
```
def forward(self, x_1):
    size = torch.ops.math.size(x_1, 0)
    size_1 = torch.ops.math.size(x_1, 1);  x_1 = None
    ones = torch.ops.aten.ones.default([1], device = device(type='cpu'), pin_memory = False)
    expand_sym_int = torch.ops.aten.expand.SymInt(ones, [size, size_1]);  ones = size = size_1 = None
    cos_default = torch.ops.aten.cos.default(expand_sym_int);  expand_sym_int = None
    return (cos_default,)
```




[ghstack-poisoned]

Chillee commented:
@pytorchbot merge
